assets/php/class-wgpb-block-grid-base.php

```diff
@@ -0,0 +1,340 @@
+		}
+
+		// Categories.
+		$this->set_categories_query_args( $query_args );
+
+		// Always query only IDs.
+		$query_args['fields'] = 'ids';
+
+		return $query_args;
+	}
+
+	/**
+	 * Set args specific to this block
+	 *
+	 * @param array $query_args Query args.
+	 */
+	abstract protected function set_block_query_args( &$query_args );
+
+	/**
+	 * Set categories query args.
+	 *
+	 * @param array $query_args Query args.
+	 */
+	protected function set_categories_query_args( &$query_args ) {
+		if ( ! empty( $this->attributes['categories'] ) ) {
+			$categories = array_map( 'absint', $this->attributes['categories'] );
+
+			$query_args['tax_query'][] = array(
+				'taxonomy'         => 'product_cat',
+				'terms'            => $categories,
+				'field'            => 'term_id',
+				'operator'         => 'all' === $this->attributes['catOperator'] ? 'AND' : 'IN',
+
+				/*
+				 * When cat_operator is AND, the children categories should be excluded,
+				 * as only products belonging to all the children categories would be selected.
+				 */
+				'include_children' => 'all' === $this->attributes['catOperator'] ? false : true,
+			);
+		}
+	}
+
+	/**
+	 * Run the query and return an array of product IDs
+	 *
+	 * @return array List of product IDs
+	 */
+	protected function get_products() {
+		// @todo Cache this
+		if ( 'product-top-rated' === $this->block_name ) {
+			add_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+			$query = new WP_Query( $this->query_args );
+			remove_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+		} else {
+			$query = new WP_Query( $this->query_args );
+		}
+
+		$results = wp_parse_id_list( $query->posts );
+
+		// Remove ordering query arguments which may have been added by get_catalog_ordering_args.
+		WC()->query->remove_ordering_args();
+
+		return $results;
+	}
+
+	/**
+	 * Render the Products block.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render() {
+		$products = $this->get_products();
+		$classes  = $this->get_classes();
+
+		$output = implode( '', array_map( array( $this, 'render_product' ), $products ) );
+
+		return sprintf( '<ul class="%s">%s</ul>', $classes, $output );
+	}
+
+	/**
+	 * Get the list of classes to apply to this block.
+	 *
+	 * @return string space-separated list of classes.
+	 */
+	public function get_classes() {
+		$classes = array(
+			'wc-block-grid',
+			"wp-block-{$this->block_name}",
+			"wc-block-{$this->block_name}",
+			"has-{$this->attributes['columns']}-columns",
+			'products',
+		);
+
+		if ( $this->attributes['rows'] > 1 ) {
+			$classes[] = 'has-multiple-rows';
+		}
+
+		if ( isset( $this->attributes['align'] ) ) {
+			$classes[] = "align{$this->attributes['align']}";
+		}
+
+		return implode( ' ', $classes );
+	}
+
+	/**
+	 * Render a single products.
+	 *
+	 * @param int $id Product ID.
+	 * @return string Rendered product output.
+	 */
+	public function render_product( $id ) {
```

Comment:
More of a question, may need some discussion.

In this method we're mixing Core template functions (such as `woocommerce_show_product_sale_flash`) with custom HTML (such as using a div for the product title instead of a the core template function which outputs the heading).

Additionally, because we're using the core template functions we're forced to use the $product global (yuk).

Do you think it's worth taking full control over the markup and avoiding core template functions? That would also ensure the block CSS we define won't be impacted by themes filtering the template HTML.

Downside, potentially more logic needed to construct the html.

Reply:
Happy to help with this if you think it's a good idea.

Reply:
@ryelle Sounds good to me. We can mirror the default HTML WooCommerce outputs, but handle it ourselves. 

Themes filtering may be affected but it will allow us to style more reliably between editor and frontend.